### ISSUE TITLE
Failed migration 0.11.1 -> 0.14.0-RC3

### ISSUE THREAD

**Issue by bobrik** [mesosphere/marathon#2984]
Turns out #2405 did not help. I tried upgrade from 0.11.1 -> 0.14.0-RC3 before, but it failed. New run fails too:

```
[2016-01-13 23:59:04,275] ERROR Failed to take over leadership (mesosphere.marathon.MarathonSchedulerService$$EnhancerByGuice$$348337d4:pool-1-thread-2)
mesosphere.marathon.MigrationFailedException: while migrating storage to 0.13
	at mesosphere.marathon.state.Migration$$anonfun$migrations$4$$anonfun$apply$2.applyOrElse(Migration.scala:61) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at mesosphere.marathon.state.Migration$$anonfun$migrations$4$$anonfun$apply$2.applyOrElse(Migration.scala:60) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at scala.util.Failure$$anonfun$recover$1.apply(Try.scala:216) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at scala.util.Try$.apply(Try.scala:192) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at scala.util.Failure.recover(Try.scala:216) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at scala.concurrent.Future$$anonfun$recover$1.apply(Future.scala:324) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at scala.concurrent.Future$$anonfun$recover$1.apply(Future.scala:324) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_66-internal]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_66-internal]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_66-internal]
Caused by: mesosphere.marathon.StoreCommandFailedException: Can not create entity framework:id
	at mesosphere.util.state.zk.ZKStore$$anonfun$exceptionTransform$1.applyOrElse(ZKStore.scala:84) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at mesosphere.util.state.zk.ZKStore$$anonfun$exceptionTransform$1.applyOrElse(ZKStore.scala:83) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	... 10 common frames omitted
Caused by: org.apache.zookeeper.KeeperException$NodeExistsException: KeeperErrorCode = NodeExists for /marathons/scan/state/framework:id
	at org.apache.zookeeper.KeeperException.create(KeeperException.java:119) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at org.apache.zookeeper.KeeperException.create(KeeperException.java:51) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at com.twitter.zk.AsyncCallbackPromise$class.process(AsyncCallbackPromise.scala:18) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at com.twitter.zk.StringCallbackPromise.process(AsyncCallbackPromise.scala:23) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at com.twitter.zk.StringCallbackPromise.processResult(AsyncCallbackPromise.scala:25) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at org.apache.zookeeper.ClientCnxn$EventThread.processEvent(ClientCnxn.java:605) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at org.apache.zookeeper.ClientCnxn$EventThread.run(ClientCnxn.java:498) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
```

`frameworkId` znode exists, but not `framework:id`.

This only happened to one of 19 marathons we have and it can be easily fixed by redeploying, but still.

**Comment by meichstedt** [comment]
Thanks for reporting this so fast, @bobrik! Turns out the migration tried to create the new framework:id node even if it existed – I could've seen this while working on #2405. Anyway, PR is on the way and a new RC in the pipeline.

**Comment by bobrik** [comment]
@meichstedt, @kolloch please make sure to include the updated mesos-ui into the new RC too.

**Comment by gkleiman** [comment]
@bobrik which updated mesos-ui? or do you mean the Marathon UI?

**Comment by bobrik** [comment]
I mean Marathon UI. I'm interested in https://github.com/mesosphere/marathon-ui/pull/549

**Comment by gkleiman** [comment]
@bobrik Marathon v0.14.0-RC4 is out, including fixes to this issue and mesosphere/marathon-ui#549.

We'd appreciate if you could try it and tell us if it fixed the migration issues.

Thanks!

**Comment by bobrik** [comment]
There are good news and there are bad news.

Good: it worked!

Bad: some time between previous migration attempts marathon got a split brain situation and `framework:id` had a different value than `frameworkId`. In fact `framework:id` value was killed in mesos. In the end I've got a new framework id associated with 0.14 and it did not see tasks launched by previous version. I killed both frameworks in mesos, removed framework id znodes and now it works.

Is it a new feature not to start up when framework is terminated in mesos? Looks like it worked before by allocating another id.

```
I0114 16:43:01.599267   117 sched.cpp:274] No credentials provided. Attempting to register without authentication
I0114 16:43:01.804908   117 sched.cpp:1027] Got error 'Framework has been removed'
I0114 16:43:01.804929   117 sched.cpp:1839] Asked to abort the driver
I0114 16:43:01.808161   117 sched.cpp:1073] Aborting framework '20150606-001827-252388362-5050-5982-0003'
I0114 16:43:01.811542   221 sched.cpp:1805] Asked to stop the driver
I0114 16:43:01.811627   139 sched.cpp:1043] Stopping framework '20150606-001827-252388362-5050-5982-0003'
```

**Comment by gkleiman** [comment]
@bobrik: thanks for testing RC4!

I am glad that the migration finally completed.

Regarding the `frameworkId`, my guess is that during one of the migration attempts, the `frameworkId` got successfully renamed to `framework:id`. Then you started an older version of Marathon, which couldn't find `frameworkId` any more and thus registered as a new framework.

I created #3004 and #3005, which should prevent this in the future.

As you noticed, Marathon v0.14.0 includes 7e019bd, which makes Marathon not start up if there's an error registering with the current framework id. We did this change in order to avoid orphaning the running tasks in case of transient errors. I'll mention this in the release notes.
